consensus/istanbul/backend/announce.go

```diff
@@ -343,52 +352,42 @@ func (sb *Backend) generateAndGossipAnnounce() error {
 		return err
 	}
 
-	// Add the generated announce message to this node's cache
-	sb.cachedAnnounceMsgsMu.Lock()
-	defer sb.cachedAnnounceMsgsMu.Unlock()
-	sb.cachedAnnounceMsgs[announceAddress] = &announceMsgCachedEntry{MsgVersion: announceVersion, MsgPayload: payload}
-
 	return sb.Multicast(nil, payload, istanbulAnnounceMsg)
 }
 
-// This function is a helper function for generateAndGossipAnnounce.  It will create the latest announce msg for this node.
-func (sb *Backend) generateAnnounce() (*istanbul.Message, uint, common.Address, error) {
+// This function is a helper function for generateAndGossipAnnounce that will
+// use the version for this node's address in the signed announce version db
+func (sb *Backend) generateAnnounce() (*istanbul.Message, error) {
 	logger := sb.logger.New("func", "generateAnnounce")
-	var enodeUrl string
-	if sb.config.Proxied {
-		if sb.proxyNode != nil {
-			enodeUrl = sb.proxyNode.externalNode.URLv4()
-		} else {
-			logger.Error("Proxied node is not connected to a proxy")
-			return nil, 0, common.Address{}, errNoProxyConnection
-		}
-	} else {
-		enodeUrl = sb.p2pserver.Self().URLv4()
-	}
 
-	// Retrieve the set of remote validators' public key to encrypt the enodeUrl
-	validatorConnSet, err := sb.retrieveValidatorConnSet()
+	enodeURL, err := sb.getEnodeURL()
 	if err != nil {
-		return nil, 0, common.Address{}, err
+		logger.Error("Error getting enode URL", "err", err)
+		return nil, err
 	}
-
-	announceRecords := make([]*announceRecord, 0, len(validatorConnSet))
-	for addr := range validatorConnSet {
-		// TODO - Need to encrypt using the remote validator's validator key
-		announceRecords = append(announceRecords, &announceRecord{DestAddress: addr, EncryptedEnodeURL: []byte(enodeUrl)})
+	announceRecords, err := sb.generateAnnounceRecords(enodeURL)
+	if err != nil {
+		logger.Warn("Error generating announce records", "err", err)
+		return nil, err
+	}
+	if len(announceRecords) == 0 {
+		logger.Trace("No announce records were generated, will not generate announce")
+		return nil, nil
+	}
+	announceVersion, err := sb.signedAnnounceVersionTable.GetVersion(sb.Address())
+	if err != nil {
+		return nil, err
 	}
 
 	announceData := &announceData{
 		AnnounceRecords: announceRecords,
-		EnodeURLHash:    istanbul.RLPHash(enodeUrl),
-		// Unix() returns a int64, but we need a uint for the golang rlp encoding implmentation. Warning: This timestamp value will be truncated in 2106.
-		Version: uint(time.Now().Unix()),
+		Version:         announceVersion,
```

Comment:
Why not the latest announce version for the sending node? The intended recipient will upsert a new entry into the val enode table with the decrypted enode URL and the version provided in the `announceData`. If the version that's sent in the message is the current timestamp it will result in a different version that's upserted in the recipient validator's val enode table from the one that it should know about via the signed announce version exchanging